### ISSUE TITLE
Support of SM2/SM3/SM4 in this library

### ISSUE THREAD

**Issue by TommyLike** [rpgp/rpgp#236]
Just cheking, if this will be supported in this pgp libaray

1. SM2: https://www.openssl.org/docs/manmaster/man7/SM2.html
2. SM3: https://www.openssl.org/docs/man1.1.1/man3/EVP_sm3.html
3. SM4: https://www.openssl.org/docs/man3.0/man7/EVP_CIPHER-SM4.html

**Comment by sorairolake** [comment]
Are they included in the OpenPGP specification? As far as I know, it doesn't seem to be in RFCs[^1][^2].

[^1]: https://datatracker.ietf.org/doc/html/rfc4880
[^2]: https://datatracker.ietf.org/doc/html/draft-ietf-openpgp-crypto-refresh

**Comment by TommyLike** [comment]
> Are they included in the OpenPGP specification? As far as I know, it doesn't seem to be in RFCs[1](#user-content-fn-1-a72789e1087e24ca6eebed70329e3e77)[2](#user-content-fn-2-a72789e1087e24ca6eebed70329e3e77).
> 
> ## Footnotes
> 1. https://datatracker.ietf.org/doc/html/rfc4880 [↩](#user-content-fnref-1-a72789e1087e24ca6eebed70329e3e77)
> 2. https://datatracker.ietf.org/doc/html/draft-ietf-openpgp-crypto-refresh [↩](#user-content-fnref-2-a72789e1087e24ca6eebed70329e3e77)

Correct, it's not in the SPEC, it's been supported in RNP and check [here](https://datatracker.ietf.org/doc/html/draft-ribose-openpgp-oscca-00)

**Comment by sorairolake** [comment]
@TommyLike Thanks for information. These Rust implementations are available,[^1][^2][^3] so this might be easy to solve.

[^1]: https://crates.io/crates/sm2
[^2]: https://crates.io/crates/sm3
[^3]: https://crates.io/crates/sm4

**Comment by dignifiedquire** [comment]
Given the availability of those crates, I'd be happy to accept a PR that adds support for these behind a feature flag.

**Comment by TommyLike** [comment]
Looking forward to someone who can help on this. 😄

**Comment by hko-s** [comment]
There are implementations of all three mechanisms in RustCrypto by now:

https://crates.io/crates/sm2/
https://crates.io/crates/sm3/
https://crates.io/crates/sm4/

However, it seems that the RustCrypto implementation of sm2 is still missing support for the SM2 Key Exchange Protocol and SM2 Public Key Encryption.

**Comment by hko-s** [comment]
While there is a draft text proposal for using these SCA cryptographic mechanisms in OpenPGP (see https://www.rnpgp.org/specs/draft-ribose-openpgp-sca/), no OpenPGP algorithm IDs have been assigned for their use. As far as I can tell, no one has signaled interest in using these algorithms in production.

(I believe the RNP OpenPGP library contains testing code for these SCA mechanisms, using algorithm codepoints that are reserved for experimental use)
